### ISSUE TITLE
How to use in ASP.NET Core?

### ISSUE THREAD

**Issue by silkfire** [Barsonax/Singularity#111]
It's not very clear from the documentation how to setup Singularity as the DI Container in an ASP.NET Core application.

This is my code in `Startup.cs`:

```
public void ConfigureServices(IServiceCollection services)
{
   services.AddControllers()
           .AddControllersAsServices()
           .SetCompatibilityVersion(CompatibilityVersion.Version_3_0);

   var container = new Container(cb =>
   {
      cb.RegisterServiceProvider();

      // My Singularity registrations
   });
}
```

Getting the generic _ASP.NET Core_ exception instead:

`InvalidOperationException: Unable to resolve service for type '<MyService>' while attempting to activate '<MyController>'.`

I'm using ASP.NET Core 3.0.

**Comment by Barsonax** [comment]
Thank you for making a issue!

The way your code is now Singularity won't do anything. You need to also resolve 'IServiceProvider' and return that in the method.

Also don't forget to call 'cb.RegisterServices' and pass so in the service collection so that all the dependencies in there will get registered.

The above should be added to the docs. Iam also considering adding the above steps to a extension method to simplify this setup.

Will check this soon.

**Comment by silkfire** [comment]
Thank you for your reply. I was able to somewhat progress with some help from this answer on SO (https://stackoverflow.com/a/58133589/633098). The setup for integrating custom IServiceProviders has changed in ASP.NET Core 3.0.

First, create a custom service provider factory:
```
internal class SingularityServiceProviderFactory : IServiceProviderFactory<Container>
{
   public Container CreateBuilder(IServiceCollection services)
   {
      var container = new Container(cb =>
      {
         cb.RegisterServiceProvider();
         cb.RegisterServices(services);
      });

      return container;
   }

   public IServiceProvider CreateServiceProvider(Container container)
   {
      return container.GetInstance<IServiceProvider>();
   }
}
```

Then register your services with Singularity in the `ConfigureContainer` method:


```
public void ConfigureContainer(ContainerBuilder builder)
{
   builder.Register<>...
}
```

But this gives me an exception `Singularity.Exceptions.CannotAutoResolveConstructorException: 'Found 2 suitable constructors for type Microsoft.Extensions.Hosting.Internal.ConsoleLifetime. Please specify the constructor explicitly.'`

**Comment by Barsonax** [comment]
Seems `Microsoft.Extensions.Hosting.Internal.ConsoleLifetime` now has 2 constructors so auto constructor selection will fail hence the error: 
Old:
https://github.com/aspnet/Hosting/blob/master/src/Microsoft.Extensions.Hosting/Internal/ConsoleLifetime.cs
New:
https://github.com/aspnet/Extensions/blob/master/src/Hosting/Hosting/src/Internal/ConsoleLifetime.cs

There is not really a robust way to do this automatically so best would be to explicitly say which constructor to pick but since this is registered in the servicecollection that will not make for very nice code so I will have to look into this further.

**Comment by silkfire** [comment]
Thanks, much appreciated.

Actually I also realized that the `public void ConfigureContainer(ContainerBuilder builder)` method won't work cause I think it expects `Container` there, but I'm not 100% sure.

**Comment by Barsonax** [comment]
Managed to get singularity working with some modifications. Its now possible to specify how to resolve the constructor (as microsoft seems to have the convention of using many multi constructor classes): https://github.com/Barsonax/Singularity/pull/112

To get it to work in asp net 3.0 mvc you have to add this in your Program.cs:
```
.UseServiceProviderFactory(new SingularityServiceProviderFactory());
```
And in Startup.cs you have to register these services:
```
services.AddSingleton<IHttpContextAccessor, HttpContextAccessor>();
services.AddSingleton<IActionContextAccessor, ActionContextAccessor>();
services.AddSingleton<IStringLocalizerFactory, ResourceManagerStringLocalizerFactory>();
```
Do this and you can launch the app and see the welcome screen.

Have to test it a bit more before I will release a new package but you can use the code in the PR if you like.

**Comment by silkfire** [comment]
Fantastic. Could you perhaps release a pre-release (preview) package instead so I can test with it?

**Comment by Barsonax** [comment]
> Fantastic. Could you perhaps release a pre-release (preview) package instead so I can test with it?

Planning to release a new pre release package today unless I find something serious.

> Also, it'd be more convenient if UseServiceProviderFactory(new SingularityServiceProviderFactory()) was abstracted away into perhaps a UseSingularity() method.

While this is certainly possible IHostBuilder is defined in `Microsoft.Extensions.Hosting.Abstractions` so that would mean adding that as a dependency to `Singularity.Microsoft.DependencyInjection`. Ofcourse I could make another `Singularity.Microsoft.Hosting` package that provides this extension method but I currently don't see enough gain as that kinda defeats the goal of adding more convenience.

> Finally, does the user really need to register these services to get it to work?

These services seem to be needed regardless of which container you use. Might also depend on if you use MVC or not. Microsoft should add these if these are always needed or provide helpers for them. In fact they do for `IHttpContextAccessor`:
```
services.AddHttpContextAccessor();
```

**Comment by silkfire** [comment]
Thanks! I'll try it out and tell you how it worked out.

Hmmm. I'm currently using Grace and there's no need to register those services at all. Check out the source code at: https://github.com/ipjohnson/Grace.DependencyInjection.Extensions.

**Comment by Barsonax** [comment]
Ok strange, sounds like some more investigation is needed. Thanks for the heads up.

**Comment by silkfire** [comment]
Btw, how do I register my services with the prerelease version?

```
public void ConfigureContainer(Container container)
{
   builder.Register<>...
}
```

won't let me access the `ContainerBuilder`...

**Comment by Barsonax** [comment]
I think this should be changed so you will have access to the container builder:
https://github.com/Barsonax/Singularity/pull/112/files#r338448950

**Comment by silkfire** [comment]
Awesome! Tbh I'm not sure how I'd do it. Returning `ContainerBuilder` from inside a closure to the outer method looks ugly.

**Comment by Barsonax** [comment]
> Returning ContainerBuilder from inside a closure to the outer method looks ugly.

Yes it is thats why I changed the Container and ContainerBuilder API a bit so it will work nicely with asp net. New prerelease package will be up soon.

**Comment by Barsonax** [comment]
New package is out: 
```
https://www.nuget.org/packages/Singularity.Microsoft.DependencyInjection/0.14.1-aspnetcore30-0016
```
Also some simple documentation on how to setup:
https://barsonax.github.io/Singularity.Docs/articles/ASP.NET%20Core%20setup.html

**Comment by Barsonax** [comment]
Made a separate issue for discussing adding a additional package for `AspNetCore.Hosting`

**Comment by silkfire** [comment]
I haven't missed your post and I really appreciate all the work you've put into this. Hopefully I'll have some time tonight to test out my setup :)

I'm curious, if it required this much refactoring to get it to work in 3.0, how much was actually functional for ASP.NET Core in 2.2?

**Comment by Barsonax** [comment]
Well it was more like aligning how Singularity does things with how Microsoft wants them. The recently added ConstructorResolvers are a good example of this. Wouldn't really call it refactoring as the code was doing what it was supposed to.

Before this issue I don't think anyone tried to use it with asp .net. I did run a small test myself a long time ago (I think that wasn't even asp net core?).

**Comment by silkfire** [comment]
Interesting. I know someone must've tried it out, because the benchmark is out there among the other ones in _[IoCPerformance](https://github.com/danielpalme/IocPerformance#advanced-features)_. 😉

**Comment by Barsonax** [comment]
Yes but that benchmark isn't running asp net core (so the naming might be a bit misleading). It just simulates what would happen in a asp net application (such as resolving scoped instances).

**Comment by silkfire** [comment]
Oh really. Yea then that's definitely misleading!

**Comment by silkfire** [comment]
I had a chance to finally test out the latest pre-release version. Unfortunately it didn't work.

If I on purpose do not register my dependencies, I get an `InvalidOperationException: Unable to resolve service for type '<MyService>' while attempting to activate '<MyController>'.` Here I expect to get an error message from Singularity, not from ASP.NET Core.

If I _do_ register my dependencies though, I get this error instead, also from ASP.NET Core:

`InvalidOperationException: No service for type 'Microsoft.AspNetCore.Mvc.Infrastructure.IActionResultExecutor`1[Microsoft.AspNetCore.Mvc.ObjectResult]' has been registered.`

**Comment by Barsonax** [comment]
> Here I expect to get an error message from Singularity, not from ASP.NET Core.

I don't think this is possible because to be compatible with Microsoft dependency injection Singularity has to return a null if it cannot resolve a dependency. This obviously throws away valuable error info but thats just the way Microsoft seems to want it.

> If I do register my dependencies though, I get another error instead, also from ASP.NET Core

This does work with other containers?

**Comment by silkfire** [comment]
You mean if try another vendor? I'm currently using Grace and it works perfectly. But I'm very keen on trying out Singularity but haven't been able to get it to work so far.

**Comment by Barsonax** [comment]
Not entirely sure why you are getting that error but maybe its because of the constructor selection thats still not optimal.

Iam thinking of adding a selector that picks the constructor in a more dynamic way by taking into account what can be resolved instead of just picking the constructor with the most or least parameters. This will allow it to be smarter but it will also postpone the constructor selector meaning you will get errors later.

Keep in mind that before this Singularity just assumed there would be 1 public constructor and was designed like that. Even if I add a smart constructor selector I will keep this the default as its the most strict and will surface possible bugs as quick a possible.

**Comment by silkfire** [comment]
Thanks, that sounds like the right way to go. Tbh I haven't really investigated how exactly Grace solves this...

**Comment by silkfire** [comment]
When you forget to register a service with Grace in ASP.NET Core, you get an exception similar to below. Note how Grace takes over the whole resolving process.

```
LocateException: Could not locate Type Wordsmith.WordFlip.Data.Repositories.IWordFlipRepository`1[Wordsmith.WordFlip.Data.Entities.FlippedSentence]
   1 Importing Wordsmith.WordFlip.WebApi.Controllers.FlipController
   2 Importing Wordsmith.WordFlip.Services.Core.WordFlippingService for constructor parameter flippingService
   3 Importing Wordsmith.WordFlip.Services.Data.WordFlipDataService for constructor parameter dataService
   4 Importing Wordsmith.WordFlip.Data.Repositories.IWordFlipRepository`1[Wordsmith.WordFlip.Data.Entities.FlippedSentence] for constructor parameter wordFlipRepository
```

Stack trace:

```
Grace.DependencyInjection.Impl.InjectionContextValueProvider.GetValueFromInjectionContext<T>(IExportLocatorScope locator, StaticInjectionContext staticContext, object key, IInjectionContext dataProvider, object defaultValue, bool useDefault, bool isRequired)
lambda_method(Closure , IExportLocatorScope , IDisposalScope , IInjectionContext )
Grace.DependencyInjection.Impl.InjectionScope.InternalLocate(IExportLocatorScope scope, IDisposalScope disposalScope, Type type, ActivationStrategyFilter consider, object key, IInjectionContext injectionContext, bool allowNull, bool isDynamic)
Grace.DependencyInjection.Impl.InjectionScope.Grace.DependencyInjection.IInjectionScope.LocateFromChildScope(IExportLocatorScope childScope, IDisposalScope disposalScope, Type type, object extraData, ActivationStrategyFilter consider, object key, bool allowNull, bool isDynamic)
Grace.DependencyInjection.Impl.ActivationStrategyDelegateCache.FallbackExecution(ImmutableHashTree<Type, ActivationStrategyDelegate> currentNode, Type type, IExportLocatorScope scope, bool allowNull, IInjectionContext context)
Grace.DependencyInjection.Impl.ActivationStrategyDelegateCache.ExecuteActivationStrategyDelegateAllowNull(Type type, IExportLocatorScope scope)
Grace.DependencyInjection.Impl.LifetimeScope.System.IServiceProvider.GetService(Type type)
Grace.AspNetCore.MVC.GraceControllerActivator.Create(ControllerContext context)
Microsoft.AspNetCore.Mvc.Controllers.ControllerFactoryProvider+<>c__DisplayClass5_0.<CreateControllerFactory>g__CreateController|0(ControllerContext controllerContext)
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, object state, bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
```

**Comment by Barsonax** [comment]
I see a 'Grace.AspNetCore.MVC.GraceControllerActivator' in the stacktrace instead of the service provider. Maybe that one is responsible for throwing the errors. Have to investigate this but that might be a very good reason for a mvc specific package.

**Comment by Barsonax** [comment]
So instead of making a custom controller activator just to make it possible for singularity to throw errors  I opted for the option of letting singularity return null when a Microsoft type is being resolved. Note iam specifically talking about resolving that type directly from the container. If its a parameter in your constructor then you will still get a nice error. 

This means that you will get singularity exceptions for all your types, not just controllers.

https://github.com/Barsonax/Singularity/tree/feature/aspnetcorefix

**Comment by Barsonax** [comment]
@silkfire I added a bestmatch constructor resolver so now Singularity should resolve services correctly in ASP.NET Core. See latest prerelease.

**Comment by silkfire** [comment]
@Barsonax Thanks for the update, appreciate all the work you put in to make this work. I hope to get some time in a few days to test the prerelease out :) I'll keep you posted.

**Comment by silkfire** [comment]
@Barsonax Apologies for the late reply, but life came in the way.

I've gotten some time to test the latest prerelease and for the first time - it works beautifully :) Well done!

I tried out the `Singularity.AspNetCore.MVC` package, adding a call to `builder.SetupMvc()` but I didn't notice any difference. What are the use cases for it?
Also, when commenting away the registration, I now get an exception from `Singularity`, just as I expected. But it's not the detailed one, with a bullet list of numbers showing the order it tried to register services. I thought that the `Singularity.AspNetCore.MVC` package might add support for that's probably just wishful thinking from my side :)

**Comment by Barsonax** [comment]
> I've gotten some time to test the latest prerelease and for the first time - it works beautifully :) Well done!

Great to hear :)

> By the way I tried out the Singularity.AspNetCore.MVC package, adding a call to builder.SetupMvc() in ConfigureContainer(ContainerBuilder builder), but I didn't notice any immediate difference. What are the use cases for it?

It takes over the creation of controllers and views from asp.net. Important for performance and better error handling.

> Also, when commenting away the service registrations, I now get an exception from Singularity, just as expected. But it's not the detailed one, with a numbered list showing the order the graph tries to register services. I thought that the Singularity.AspNetCore.MVC package might add support for that, but that's just probably wishful thinking from my side :)

There was a bug in the error handling which is now fixed. Look in the inner exceptions for details why a service failed to resolve.

Now that asp.net core seems to be working properly I released a new version 0.15 so you won't have to use the prerelease anymore.

**Comment by Barsonax** [comment]
Closing this as the issue was resolved in v0.15
